### PR TITLE
Full cluster backup/restore

### PR DESCRIPTION
**Pull request by titanous**
`flynn cluster backup` saves a full-cluster backup to a local tar file. The backup consists of expanded app details for the discoverd, flannel, postgres, and controller apps and a SQL dump of all databases in the postgres cluster.

`flynn-host bootstrap --from-backup` bootstraps a cluster and restores a backup. First the discoverd, flannel, and postgres apps are started on the cluster, then the SQL is restored along with some extra commands that update the artifact URLs to the latest ones from the bootstrap manifest.

After the data is restored, the controller/scheduler is started and the scheduler handles starting the rest of the Flynn and user apps.

Refs #1244

/cc @temujin9

### CODE REVIEW

File: cli/cluster.go [hunk 1]
```diff
@@ -268,3 +296,82 @@ func runClusterMigrateDomain(args *docopt.Args) error {
...
 }
+
+func runClusterBackup(args *docopt.Args) error {
+	client, err := getClusterClient()
+	if err != nil {
+		return err
+	}
+
+	var bar *pb.ProgressBar
+	if !args.Bool["--quiet"] && term.IsTerminal(os.Stderr.Fd()) {
```
Comment (lmars) [inline cli/cluster.go]: `--quiet` isn't in the usage options.

File: cli/cluster.go [hunk 2]
```diff
@@ -268,3 +296,82 @@ func runClusterMigrateDomain(args *docopt.Args) error {
...
+		if err != nil {
+			return fmt.Errorf("error getting %s app release: %s", name, err)
+		}
+		formation, err := client.GetFormation(app.ID, release.ID)
+		if err != nil {
+			return fmt.Errorf("error getting %s app formation: %s", name, err)
+		}
+		artifact, err := client.GetArtifact(release.ArtifactID)
+		if err != nil {
+			return fmt.Errorf("error getting %s app release: %s", name, err)
```
Comment (lmars) [inline cli/cluster.go]: s/release/artifact

File: host/cli/bootstrap.go [hunk 3]
```diff
@@ -88,13 +97,225 @@ func runBootstrap(args *docopt.Args) {
...
 	if err != nil {
 		os.Exit(1)
 	}
 }
 
+func runBootstrapBackup(manifest []byte, backupFile string, ch chan *bootstrap.StepInfo, clusterURL string, hostIPs []string, minHosts, timeout int) error {
+	defer close(ch)
+	f, err := os.Open(backupFile)
+	if err != nil {
+		log.Fatalf("error opening backup file: %s", err)
```
Comment (lmars) [inline host/cli/bootstrap.go]: Should this not be returning `fmt.Errorf`?

**Comment by lmars** [comment]
LGTM pending minor comments

File: cli/cluster.go [hunk 4]
```diff
@@ -268,3 +296,82 @@ func runClusterMigrateDomain(args *docopt.Args) error {
...
 	}
 }
+
+func runClusterBackup(args *docopt.Args) error {
+	client, err := getClusterClient()
+	if err != nil {
+		return err
+	}
+
+	var bar *pb.ProgressBar
```
Comment (lmars) [inline cli/cluster.go]: Would be nice to see a log message before / after the progress bar, e.g.:

```
$ flynn cluster backup --file ~/flynn.backup
Creating cluster backup...
70.72 MB 6.51 MB/s 10s
Backup complete
```

Comment (titanous) [inline cli/cluster.go]: Done.

**Comment by lmars** [comment]
Successfully tested a cluster backup / restore with an app that uses a postgres resource.

LGTM :smile:

**Comment by kwando** [comment]
Is the cluster certificate backed up and restored?

**Comment by titanous** [comment]
Yes _everything_ is backed and restored.

**Pull request merged**